File: packet_seaquence/log/python_no_shark/src/protocol_analyzer/sctp.py
from scapy.all import SCTP
from src.protocol_analyzer.base import ProtocolAnalyzer

class AnalyzeSctp(ProtocolAnalyzer):
    """
    SCTP内の詳細情報を解析するクラス
    """

    def analyze(self):
        if SCTP not in self.packet:
            return None
        try:
            sctp_info = {}
            sctp_layer = self.packet[SCTP]
            
            # 送信元ポート
            sctp_info['src_port'] = str(sctp_layer.sport)
            
            # 宛先ポート
            sctp_info['dst_port'] = str(sctp_layer.dport)
            
            # 検証タグ
            if hasattr(sctp_layer, 'tag'):
                sctp_info['verification_tag'] = str(sctp_layer.tag)
                
            # チェックサム
            if hasattr(sctp_layer, 'chksum'):
                sctp_info['checksum'] = hex(sctp_layer.chksum) if sctp_layer.chksum is not None else "None"
            
            return sctp_info
            
        except Exception as e:
            print(f"SCTP解析中にエラーが発生しました: {str(e)}")
            import traceback
            traceback.print_exc()
            return None

    def get_display_info(self):
        info = self.analyze()
        if not info:
            return "SCTP情報なし"
        return f"SCTP {info.get('src_port','?')} -> {info.get('dst_port','?')} (VTAG={info.get('verification_tag','?')})"

    def get_summary(self):
        info = self.analyze()
        if not info:
            return "SCTP"
        return f"SCTP {info.get('src_port','?')} -> {info.get('dst_port','?')}"

File: packet_seaquence/log/go_src/protocol_analyzer/sctp.go
// Package protocol_analyzer provides SCTP protocol analyzer
package protocol_analyzer

import (
	"fmt"
)

// AnalyzeSCTP implements the ProtocolAnalyzer interface for SCTP
type AnalyzeSCTP struct {
	Pkt Packet
}

// NewAnalyzeSCTP creates a new SCTP analyzer
func NewAnalyzeSCTP(packet Packet) *AnalyzeSCTP {
	return &AnalyzeSCTP{
		Pkt: packet,
	}
}

// Analyze extracts SCTP information from a packet
func (a *AnalyzeSCTP) Analyze() map[string]interface{} {
	if !a.Pkt.LayerExists("sctp") {
		return nil
	}

	sctpInfo := make(map[string]interface{})
	
	// Source port
	if srcPort, ok := a.Pkt.GetField("sctp", "srcport"); ok {
		sctpInfo["src_port"] = srcPort
	}
	
	// Destination port
	if dstPort, ok := a.Pkt.GetField("sctp", "dstport"); ok {
		sctpInfo["dst_port"] = dstPort
	}
	
	// Verification tag
	if verificationTag, ok := a.Pkt.GetField("sctp", "verification_tag"); ok {
		sctpInfo["verification_tag"] = verificationTag
	}
	
	return sctpInfo
}

// GetDisplayInfo returns formatted SCTP information
func (a *AnalyzeSCTP) GetDisplayInfo() string {
	info := a.Analyze()
	if info == nil {
		return "SCTP情報なし"
	}
	
	srcPort, _ := info["src_port"].(string)
	if srcPort == "" {
		srcPort = "?"
	}
	
	dstPort, _ := info["dst_port"].(string)
	if dstPort == "" {
		dstPort = "?"
	}
	
	verificationTag, _ := info["verification_tag"].(string)
	if verificationTag == "" {
		verificationTag = "?"
	}
	
	return fmt.Sprintf("SCTP %s -> %s (VTAG=%s)", srcPort, dstPort, verificationTag)
}

// GetSummary returns a summary of SCTP information
func (a *AnalyzeSCTP) GetSummary() string {
	info := a.Analyze()
	if info == nil {
		return "SCTP"
	}
	
	srcPort, _ := info["src_port"].(string)
	if srcPort == "" {
		srcPort = "?"
	}
	
	dstPort, _ := info["dst_port"].(string)
	if dstPort == "" {
		dstPort = "?"
	}
	
	return fmt.Sprintf("SCTP %s -> %s", srcPort, dstPort)
}